src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,164 @@
+    {
+        $this->ctx = hash_init('md5');
+        $uploadDate = time();
+        $objectId = new \MongoDB\BSON\ObjectId();
+        $main_file = [
+            "chunkSize" => $chunkSizeBytes,
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($filesCollection, $chunksCollection, $chunkSizeBytes);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        while ($data = fread($source, $this->chunkSizeBytes)) {
+            $this->insertChunk($data);
+        }
+        return $this->fileCollectionInsert();
+    }
+    /**
+    * Insert a chunks into GridFS from a string
+    *
+    * @param string  $toWrite   String to upload
+    * @return int
+    */
+    public function insertChunks($toWrite)
+    {
+        rewind($this->buffer);
+        $cached = fread($this->buffer, $this->chunkSizeBytes);
+        $toWrite = $cached . $toWrite;
+        $bytes = str_split($toWrite, $this->chunkSizeBytes);
+
+        fclose($this->buffer);
+        $this->buffer = fopen("php://temp", "w+");
+
+        for ($i = 0; $i < count($bytes); $i++) {
+            if ($i == count($bytes) - 1 && strlen($bytes[$i]) < $this->chunkSizeBytes) {
+                fwrite($this->buffer, $bytes[$i]);
+                return strlen($toWrite);
+            }
+            $this->insertChunk($bytes[$i]);
+        }
+        return strlen($toWrite);
+    }
+    /**
+    * Close an active stream, pushes all buffered data to GridFS
+    *
+    */
+    public function close()
+    {
+        rewind($this->buffer);
+        $cached = fread($this->buffer, $this->chunkSizeBytes);
+
+        if(strlen($cached) > 0) {
```

Comment:
If the buffer always points to its end, you may be able to use [`ftell()`](http://php.net/ftell) on it instead of loading it into a string to run `strlen()`. This may be a moot point, though, since we either read nothing or need to access it anyway to insert the chunk.